### ISSUE TITLE
Syntactic sugar for piping, redirection

### ISSUE THREAD

**Issue by FranklinChen** [Gabriel439/Haskell-Turtle-Library#28]
I know this is part of a much larger topic of syntax, but I am a fan of the reverse function application "pipe" operator extremely popular these days, in the OCaml and Elixir and Clojure communities, for example. I personally define it for myself in Haskell as

```haskell
infixl 0 |>
(|>) :: a -> (a -> b) -> b
x |> f = f x
```
Using this syntax, some things in Turtle look much more accessible:

```haskell
inshell "ls" empty |> stdout

stdin |> stdout

("Test" <|> "ABC" <|> "42") |> output "file.txt"

input "file.txt" |> stdout

input "file.txt" |> grep "ABC" |> stdout
```

Any interest in promoting this syntax? I believe it would be a huge win for those coming from Unix and basically anyone coming from an OO "topic-comment" or "noun-verb" way of thinking.

**Comment by Gabriel439** [comment]
This is a gray area for me.

One the one hand I understand the appeal because one complaint I get from beginners about Haskell in general is that it reads from right-to-left instead of left-to-right.  I expect this complaint to be magnified for `turtle` since Unix pipelines also read from left-to-right.

On the other hand, I try to avoid non-associative operators.

However, I'll make a compromise: if you can get this operator into `base` then I'll re-export it in `turtle`.

**Comment by Gabriel439** [comment]
Also, let me clarify that not only would I re-export it, I would also backport it for earlier versions of `base` that don't have it (using `CPP` `ifdef`s).

**Comment by FranklinChen** [comment]
"if you can get this operator into `base`": ha, great response!

**Comment by michaelt** [comment]
For what it may be worth, `(&) :: a -> (a -> b) -> b` is now in `Data.Function`, to judge from my ghc-7.10-rc1 or whatever it is.

**Comment by Gabriel439** [comment]
@FranklinChen Then would you be okay with `turtle` back-porting `(&)` to earlier versions of `base`?

**Comment by FranklinChen** [comment]
So I see `(&)` in https://github.com/ghc/ghc/blob/master/libraries/base/Data/Function.hs but it's not re-exported into the Prelude https://github.com/ghc/ghc/blob/master/libraries/base/Prelude.hs, sadly.

Not my personal choice of operator (especially in the context of Unix's use of ampersand to mean backgrounding), but if it's what the community has decided to use moving forward, I will follow suit.

**Comment by michaelt** [comment]
I agree the F# `|>` would have better suited this library (it also fits with `>>>` composition nicely, if I remember the usual arguments); but `&` proved a winner in the lens libraries, I think not without reason.  It certainly beats `$`.   But one is going to have to get used to new things with this or any library -- I just made my second attempt comprehending it last night. So the main question you raise is whether the order aids comprehension, and the sign would be worth exporting and maybe throwing around in the tutorial a bit. In fact, I was finding it pleasantly readable, especially if I added space around it to make a bit of caesura

       input "a.txt" <|> input "b.txt"  &  grep (has "4")  &  sed exclaim  &  output "c.txt"

(Pardon arbitrary expression.)  Obviously unix symbol associations aren't going to help here, but the associated image of 'flow' was seeming not too bad.  One might want, e.g. a flipped `fold` for this style.
